src/include/server/mir/scene/surface_observer.h

```diff
@@ -62,8 +62,14 @@ class SurfaceObserver
     virtual void input_consumed(Surface const* surf, std::shared_ptr<MirEvent const> const& event) = 0;
     virtual void application_id_set_to(Surface const* surf, std::string const& application_id) = 0;
     virtual void depth_layer_set_to(Surface const* surf, MirDepthLayer depth_layer) = 0;
-    virtual void entered_output(Surface const* surf, graphics::DisplayConfigurationOutputId const& id) = 0;
-    virtual void left_output(Surface const* surf, graphics::DisplayConfigurationOutputId const& id) = 0;
+    virtual void entered_output(
+        Surface const* surf,
+        graphics::DisplayConfigurationOutputId const& id,
+        std::function<bool(graphics::DisplayConfigurationOutputId const&)> const& hook) = 0;
+    virtual void left_output(
+        Surface const* surf,
+        graphics::DisplayConfigurationOutputId const& id,
+        std::function<bool(graphics::DisplayConfigurationOutputId const&)> const& hook) = 0;
```

Comment:
Adding this parameter to this interface looks very odd. Why would a generic Surface observer want to be passed a boolean functor that takes one of the other parameters? And what are the semantics of it?

Reply:
Additionally, the hook specifically addresses cases where an output is enabled at runtime. Without the hook, the scene surface might assume it has entered an output, but the frontend never emitted the event because the output instance wasn't yet ready when the frontend handler processed it. When an output is dynamically enabled, it may take some time for the new `wl_resource` to bind to the output instance. Only late `entered_output` notifications will catch up with the new output instance.

Reply:
All of which discussion confirms that responsibilities are not being encapsulated. This interaction is not "Observer" pattern: What happens if there are multiple observers, how do they co-ordinate updating the scene through "hook"?

The scene should notify observers when an enter occurs, it is not be responsible for tracking whether the frontend has actually sent a notification. Likewise the frontend should not be responsible for calling a callback to update the scene.

I take the point about the frontend being unable to send `enter` before the output is bound, but wouldn't that be better handled by recording "pending enter" in the frontend? And sending that if/when the output is bound.

Reply:
I agree. I'll revert to the previous tracking approach and start recording the pending enter events in the frontend. Addressing the missing enter events was the only reason to start using hooks.

Reply:
Done.